### PR TITLE
feat(react-native-payments): request only contact fields asked for by the user

### LINKED ISSUE #228: Allow for optional requiredShippingContactFields
As we know from [the docs](https://developer.apple.com/documentation/passkit_apple_pay_and_wallet/pkpaymentrequest/1619221-billingcontact) it's only possible to get back the postalAddress from the `requiredBillingContactFields`. If you want any of the other fields, such as name, email, and phone, you need to get them back from the `requiredShippingContactFields`. 

The issue is that I don't want to ask the user for the shipping address because they won't be shipped anything. I do, however, want their contact information. This can be achieved on the native side by removing `PKContactFieldPostalAddress` from `requiredShippingContactFields`. 

There's a few ways the js api can be extended to support this. 
1.  Split the current `requestShipping` into `requestShippingContactInfo` and `requestShippingAddress`. This would be a breaking change. 
2. Do the above but keep `requestShipping` for backwards compatibility which sets both `requestShippingContactInfo` and `requestShippingAddress` to true. No breaking change. 
3. Keep `requestShipping` and add `noShippingAddress`. No breaking change. 
4. Make `requestShipping` an object where you can specify which fields you want. Breaking change but more future proof.

Let me know your thoughts. I'd be happy to submit a PR with whatever path makes sense. 
Thank you!

### PR DESCRIPTION
**Pull request by KrastanD**
This PR resolves #228.

I've added an example and tested it on both android and ios. 

At a high level, I added new fields `requestPayerName`, `requestPayerPhone`, `requestPayerEmail` and renamed `requestBilling` to `requestBillingAddress` [in accordance with W3C](https://www.w3.org/TR/payment-request/#paymentoptions-dictionary). 

On iOS, I changed `requiredBillingContactFields` and `requiredShippingContactFields` to arrays of enum `PKContactField` [in accordance with PassKit](https://developer.apple.com/documentation/passkit_apple_pay_and_wallet/pkpaymentrequest/2865927-requiredshippingcontactfields?language=objc). On the native side I converted from the arrays from strings to PKContactFields and passed those to the paymentRequest.

On Android, I asked for data only if it was specifically requested by the user with the new fields. The only caveat is that if `requestPayerPhone`, `requestBillingAddress`, or `requestShipping` is set, `payerName` will be returned even if it wasn't requested. This is a consequence of [all addresses from google pay returning a name](https://developers.google.com/pay/api/android/reference/response-objects#PaymentData).

### CODE REVIEW

**Comment by vitalyiegorov** [comment]
@KrastanD please fix conflicts

**Comment by KrastanD** [comment]
@vitalyiegorov would you mind taking another look at this PR?

**Comment by vitalyiegorov** [comment]
@KrastanD Lets have a look on the [Stripe](https://docs.stripe.com/js/payment_request/create#stripe_payment_request-options-requestPayerName) for solving similar problem, 

# requestPayerName
By default, the browser‘s payment interface only asks the customer for actual payment information. A customer name can be collected by setting this option to true. This collected name will appears in the [PaymentResponse](https://docs.stripe.com/js/appendix/payment_response) object.

We highly recommend you collect name as this also results in collection of billing address for Apple Pay. The billing address can be used to perform address verification and block fraudulent payments. For all other payment methods, the billing address is automatically collected when available.

# requestPayerEmail
# requestPayerPhone

What are you thoughts on this?

**Comment by KrastanD** [comment]
@vitalyiegorov From how I'm understanding it, Stripe don't have a requestBillingAddress but instead automatically request it on everything except ApplePay where they require the user to requestName instead. 

This goes against the W3C Spec and I think is a more confusing API.

File: packages/react-native-payments/ios/Payments.mm [hunk 1]
```diff
@@ -377,6 +400,23 @@ - (PKMerchantCapability)merchantCapabilityFromString:(NSString *)capabilityStrin
...
+    NSDictionary<NSString *, PKContactField> *contactFieldMapping = @{
+        @"PKContactFieldName": PKContactFieldName,
+        @"PKContactFieldPostalAddress": PKContactFieldPostalAddress,
+        @"PKContactFieldEmailAddress": PKContactFieldEmailAddress,
+        @"PKContactFieldPhoneNumber": PKContactFieldPhoneNumber,
+    };
+
+    PKContactField field = contactFieldMapping[inputString];
+    
+    if (field != nil) {
```
Comment (vitalyiegorov) [inline packages/react-native-payments/ios/Payments.mm:413]: why do we need this condition? will `return field ` do the same?

File: packages/react-native-payments-example/src/app/app.tsx [hunk 2]
```diff
@@ -56,6 +58,21 @@ export const App = (): JSX.Element => {
         setTimeout(() => void paymentRequest.abort(), 1000);
     };
 
+    const handlePayWithoutShipping = (): void => {
```
Comment (vitalyiegorov) [inline packages/react-native-payments-example/src/app/app.tsx:61]: Maybe lets improve the example where you can set new fields with toggles for each type of the requested data: 
- requestBillingAddress
- requestPayerEmail
- requestPayerName
- requestPayerPhone
- requestShipping

and then we do not need separate methods/buttons?

File: packages/react-native-payments-example/src/createPaymentRequest.ts [hunk 3]
```diff
@@ -0,0 +1,26 @@
+import {
```
Comment (vitalyiegorov) [inline packages/react-native-payments-example/src/createPaymentRequest.ts:1]: please rename the file in the `snake-case` format

File: packages/react-native-payments/ios/Payments.mm [hunk 4]
```diff
@@ -69,6 +69,29 @@ - (dispatch_queue_t)methodQueue
         }
     }
 
+    NSMutableArray *requiredBillingContactFields = [NSMutableArray array];
```
Comment (vitalyiegorov) [inline packages/react-native-payments/ios/Payments.mm:72]: Do we need to have the same implementation for the Android?

File: packages/react-native-payments/src/class/payment-request/payment-request.ts [hunk 5]
```diff
@@ -249,6 +255,12 @@ export class PaymentRequest {
             IosPKMerchantCapability.PKMerchantCapabilityCredit,
         ];
 
+        const requiredBillingFields = this.createArrayOfRequestedBillingFields(methodData);
```
Comment (vitalyiegorov) [inline packages/react-native-payments/src/class/payment-request/payment-request.ts:258]: inline this calls as we use this `const` only once

**Review by vitalyiegorov — CHANGES_REQUESTED** [review]
Good job, after reading through the https://www.w3.org/TR/payment-request/#paymentoptions-dictionary and stripe docs, i agree with this implementation, some minor things left for the review,

Please update the docs/readme for according to this changes

File: packages/react-native-payments/src/class/payment-request/payment-request.ts [hunk 6]
```diff
@@ -257,8 +269,37 @@ export class PaymentRequest {
...
                 : defaultMerchantCapabilities,
-            ...(methodData.requestBilling === true && { requiredBillingContactFields: true }),
-            ...(methodData.requestShipping === true && { requiredShippingContactFields: true }),
+            ...(methodData.requestBillingAddress === true && { requiredBillingContactFields: requiredBillingFields }),
+            ...(isShippingRequired && { requiredShippingContactFields: requiredShippingFields }),
         };
     }
+
+    // eslint-disable-next-line class-methods-use-this,@typescript-eslint/class-methods-use-this
+    private createArrayOfRequestedBillingFields(methodData: IosPaymentMethodDataDataInterface): IOSPKContactField[] {
```
Comment (vitalyiegorov) [inline packages/react-native-payments/src/class/payment-request/payment-request.ts:278]: `getRequestedBillingFields`

File: packages/react-native-payments/src/class/payment-request/payment-request.ts [hunk 7]
```diff
@@ -257,8 +269,37 @@ export class PaymentRequest {
...
+        const requiredBillingFields = [];
+        if (methodData.requestBillingAddress ?? false) {
+            requiredBillingFields.push(IOSPKContactField.PKContactFieldPostalAddress);
+        }
+
+        return requiredBillingFields;
+    }
+
+    // eslint-disable-next-line class-methods-use-this,@typescript-eslint/class-methods-use-this
+    private createArrayOfRequestedShippingFields(methodData: IosPaymentMethodDataDataInterface): IOSPKContactField[] {
```
Comment (vitalyiegorov) [inline packages/react-native-payments/src/class/payment-request/payment-request.ts:288]: `getRequestedShippingFields`

File: packages/react-native-payments/ios/Payments.mm (hunk 4 above)
Comment (KrastanD) [reply packages/react-native-payments/ios/Payments.mm:72]: We don't. We're able to do it all from the js side on Android. The changes in `packages/react-native-payments/src/class/payment-request/payment-request.ts` are all that's needed.

**Review by vitalyiegorov — CHANGES_REQUESTED** [review]
@KrastanD Great job! Can you please update the readme?

**Comment by KrastanD** [comment]
@vitalyiegorov Sorry for the delay!

**Review by vitalyiegorov — APPROVED** [review]

**Pull request merged**